<!doctype html>
<html class="no-js" lang="en" data-content_root="./">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="viewport" content="width=device-width, initial-scale=1" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="iss_preprocess.reg package" href="iss_preprocess.reg.html" /><link rel="prev" title="iss_preprocess.io package" href="iss_preprocess.io.html" />

    <!-- Generated with Sphinx 8.1.3 and Furo 2024.08.06 -->
        <title>iss_preprocess.pipeline package - iss-preprocess 0.2.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?v=354aac6f" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?v=302659d7" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-with-moon" viewBox="0 0 24 24">
    <title>Auto light/dark, in light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path style="opacity: 50%" d="M 5.411 14.504 C 5.471 14.504 5.532 14.504 5.591 14.504 C 3.639 16.319 4.383 19.569 6.931 20.352 C 7.693 20.586 8.512 20.551 9.25 20.252 C 8.023 23.207 4.056 23.725 2.11 21.184 C 0.166 18.642 1.702 14.949 4.874 14.536 C 5.051 14.512 5.231 14.5 5.411 14.5 L 5.411 14.504 Z"/>
      <line x1="14.5" y1="3.25" x2="14.5" y2="1.25"/>
      <line x1="14.5" y1="15.85" x2="14.5" y2="17.85"/>
      <line x1="10.044" y1="5.094" x2="8.63" y2="3.68"/>
      <line x1="19" y1="14.05" x2="20.414" y2="15.464"/>
      <line x1="8.2" y1="9.55" x2="6.2" y2="9.55"/>
      <line x1="20.8" y1="9.55" x2="22.8" y2="9.55"/>
      <line x1="10.044" y1="14.006" x2="8.63" y2="15.42"/>
      <line x1="19" y1="5.05" x2="20.414" y2="3.636"/>
      <circle cx="14.5" cy="9.55" r="3.6"/>
    </svg>
  </symbol>
  <symbol id="svg-moon-with-sun" viewBox="0 0 24 24">
    <title>Auto light/dark, in dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round"
      class="icon-custom-derived-from-feather-sun-and-tabler-moon">
      <path d="M 8.282 7.007 C 8.385 7.007 8.494 7.007 8.595 7.007 C 5.18 10.184 6.481 15.869 10.942 17.24 C 12.275 17.648 13.706 17.589 15 17.066 C 12.851 22.236 5.91 23.143 2.505 18.696 C -0.897 14.249 1.791 7.786 7.342 7.063 C 7.652 7.021 7.965 7 8.282 7 L 8.282 7.007 Z"/>
      <line style="opacity: 50%" x1="18" y1="3.705" x2="18" y2="2.5"/>
      <line style="opacity: 50%" x1="18" y1="11.295" x2="18" y2="12.5"/>
      <line style="opacity: 50%" x1="15.316" y1="4.816" x2="14.464" y2="3.964"/>
      <line style="opacity: 50%" x1="20.711" y1="10.212" x2="21.563" y2="11.063"/>
      <line style="opacity: 50%" x1="14.205" y1="7.5" x2="13.001" y2="7.5"/>
      <line style="opacity: 50%" x1="21.795" y1="7.5" x2="23" y2="7.5"/>
      <line style="opacity: 50%" x1="15.316" y1="10.184" x2="14.464" y2="11.036"/>
      <line style="opacity: 50%" x1="20.711" y1="4.789" x2="21.563" y2="3.937"/>
      <circle style="opacity: 50%" cx="18" cy="7.5" r="2.169"/>
    </svg>
  </symbol>
  <symbol id="svg-pencil" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-pencil-code">
      <path d="M4 20h4l10.5 -10.5a2.828 2.828 0 1 0 -4 -4l-10.5 10.5v4" />
      <path d="M13.5 6.5l4 4" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
  <symbol id="svg-eye" viewBox="0 0 24 24">
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-eye-code">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M10 12a2 2 0 1 0 4 0a2 2 0 0 0 -4 0" />
      <path
        d="M11.11 17.958c-3.209 -.307 -5.91 -2.293 -8.11 -5.958c2.4 -4 5.4 -6 9 -6c3.6 0 6.6 2 9 6c-.21 .352 -.427 .688 -.647 1.008" />
      <path d="M20 21l2 -2l-2 -2" />
      <path d="M17 17l-2 2l2 2" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>

<a class="skip-to-content muted-link" href="#furo-main-content">Skip to content</a>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">iss-preprocess 0.2.1 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
          <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">iss-preprocess 0.2.1 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="getting_started.html">Getting started</a></li>
<li class="toctree-l1"><a class="reference internal" href="project_and_average.html">Crunch the data</a></li>
<li class="toctree-l1"><a class="reference internal" href="registration.html">Registration</a></li>
<li class="toctree-l1"><a class="reference internal" href="call.html">Calling spots</a></li>
<li class="toctree-l1"><a class="reference internal" href="correction.html">Channel brightness and illumination correction</a></li>
<li class="toctree-l1"><a class="reference internal" href="bleedthrough.html">Bleedthrough matrix estimation</a></li>
<li class="toctree-l1"><a class="reference internal" href="sequencing.html">Calling various spots</a></li>
<li class="toctree-l1"><a class="reference internal" href="hybridisation.html">Detecting hybridisation spots</a></li>
<li class="toctree-l1"><a class="reference internal" href="segmentation.html">Cell segmentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="detect_mcherry.html">Detection mCherry cells</a></li>
</ul>
<p class="caption" role="heading"><span class="caption-text">API reference:</span></p>
<ul class="current">
<li class="toctree-l1 current has-children"><a class="reference internal" href="iss_preprocess.html">iss_preprocess package</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" role="switch" type="checkbox"/><label for="toctree-checkbox-1"><div class="visually-hidden">Toggle navigation of iss_preprocess package</div><i class="icon"><svg><use href="#svg-arrow-right"></use></svg></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.call.html">iss_preprocess.call package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.config.html">iss_preprocess.config package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.coppafish.html">iss_preprocess.coppafish package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.diagnostics.html">iss_preprocess.diagnostics package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.image.html">iss_preprocess.image package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.io.html">iss_preprocess.io package</a></li>
<li class="toctree-l2 current current-page"><a class="current reference internal" href="#">iss_preprocess.pipeline package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.reg.html">iss_preprocess.reg package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.segment.html">iss_preprocess.segment package</a></li>
<li class="toctree-l2"><a class="reference internal" href="iss_preprocess.vis.html">iss_preprocess.vis package</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="flowchart_legends.html">Flowcharts legends</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          <div class="view-this-page">
  <a class="muted-link" href="_sources/iss_preprocess.pipeline.rst.txt" title="View this page">
    <svg><use href="#svg-eye"></use></svg>
    <span class="visually-hidden">View this page</span>
  </a>
</div>
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto-light"><use href="#svg-sun-with-moon"></use></svg>
              <svg class="theme-icon-when-auto-dark"><use href="#svg-moon-with-sun"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main" id="furo-main-content">
          <section id="iss-preprocess-pipeline-package">
<h1>iss_preprocess.pipeline package<a class="headerlink" href="#iss-preprocess-pipeline-package" title="Link to this heading">¶</a></h1>
<section id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Link to this heading">¶</a></h2>
</section>
<section id="module-iss_preprocess.pipeline.ara_registration">
<span id="iss-preprocess-pipeline-ara-registration-module"></span><h2>iss_preprocess.pipeline.ara_registration module<a class="headerlink" href="#module-iss_preprocess.pipeline.ara_registration" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.check_reg">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">check_reg</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_folder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.check_reg" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.crop_overview_registration">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">crop_overview_registration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overview_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DAPI_1_1'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.crop_overview_registration" title="Link to this definition">¶</a></dt>
<dd><p>Crop the registered overview to the same size as the reference</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>rois</strong> (<em>list</em><em>, </em><em>optional</em>) – List of rois to crop. Defaults to None.</p></li>
<li><p><strong>overview_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the overview image. Defaults to
“DAPI_1_1”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of cropped images</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>imgs (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.find_roi_position_on_cryostat">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">find_roi_position_on_cryostat</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.find_roi_position_on_cryostat" title="Link to this definition">¶</a></dt>
<dd><p>Find the A/P position of each ROI relative to the first collected slice</p>
<p>The section order is guess from the sign of <cite>section_thickness_um</cite>, positive for
antero-posterior slicing (starting from the olfactory bulb), negative for opposite.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>For each ROI, the slice depth in um relative to the</dt><dd><p>first collected slice</p>
</dd>
</dl>
<p>min_step (float): Minimum thickness between two slices</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>roi_slice_pos_um (dict)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.load_coordinate_image">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">load_coordinate_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">registered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.load_coordinate_image" title="Link to this definition">¶</a></dt>
<dd><p>Load the 3 channel image of ARA coordinates for <cite>roi</cite></p>
<p>The reference atlas is first registered to a downsampled version of the overview,
this is then registered to the normal acquisition. The coordinates of the overview
can be loaded with <cite>registered=False</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Number of the ROI</p></li>
<li><p><strong>full_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, returns the full scale image, otherwise
the downsample version used for registration. Defaults to False.</p></li>
<li><p><strong>registered</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, load the registered coordinates, otherwise
the coordinates of the overview, before shifting/cropping. Defaults to True.</p></li>
<li><p><strong>return_fname</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return the filename of the image.
Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>3 channel image of ARA coordinates</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>coords (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.load_registration_reference">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">load_registration_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.load_registration_reference" title="Link to this definition">¶</a></dt>
<dd><p>Load the registration reference image of one ROI</p>
<p>This is the downsampled version of the overview image used for registration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Number of the ROI</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Registration reference image</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>ref (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.load_registration_reference_metadata">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">load_registration_reference_metadata</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.load_registration_reference_metadata" title="Link to this definition">¶</a></dt>
<dd><p>Load metadata file associated with registration reference of one ROI</p>
<p>This is the “registration_reference_r{roi}_sl{slice_number}.yml” file that contains
shape and downsampling info.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Number of the roi</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Content of the metadata yml file</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>metadata (dict)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.make_area_image">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">make_area_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atlas_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">registered</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.make_area_image" title="Link to this definition">¶</a></dt>
<dd><p>Generate an image with area ID in each pixel</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Roi number to generate</p></li>
<li><p><strong>atlas_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Pixel size of the atlas used to find area id.
Defaults to 10.</p></li>
<li><p><strong>full_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, returns the full scale image, otherwise
the downsample version used for registration. Defaults to False.</p></li>
<li><p><strong>reload</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, reload the area image, otherwise recompute it.
Valid only if full_scale is False. Defaults to True.</p></li>
<li><p><strong>registered</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, load the registered coordinates, otherwise
the coordinates of the overview, before shifting/cropping. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Image with area id of each pixel</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>area_id (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.overview_single_roi">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">overview_single_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slice_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">chan2use=(0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">3)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigma_blur=10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">agg_func=&lt;function</span> <span class="pre">nanmean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prefix='genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subresolutions=5</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_pixel_size=2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_similar_overview=False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.overview_single_roi" title="Link to this definition">¶</a></dt>
<dd><p>Stitch and save a single ROI overview for use in atlas registration</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Number of the ROI</p></li>
<li><p><strong>slice_id</strong> (<em>int</em>) – Slice number to stitch</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the acquisition to plot.</p></li>
<li><p><strong>chan2use</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Channels to use for stitching. Defaults to
(0, 1, 2, 3).</p></li>
<li><p><strong>sigma_blur</strong> (<em>int</em><em>, </em><em>optional</em>) – Sigma for gaussian blur. Defaults to 10.</p></li>
<li><p><strong>agg_func</strong> (<em>function</em><em>, </em><em>optional</em>) – Aggregation function to apply across channels.
Defaults to np.nanmean. Unused if <cite>non_similar_overview</cite> is True.</p></li>
<li><p><strong>ref_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the reference image. Defaults to
“genes_round”.</p></li>
<li><p><strong>subresolutions</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of subresolutions to save. Defaults to 5.</p></li>
<li><p><strong>max_pixel_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum pixel size for the pyramid. Defaults to
2.</p></li>
<li><p><strong>non_similar_overview</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, stitch the overview tiles with</p></li>
<li><p><strong>by</strong> (<em>the stitch_tiles function rather than stitch_registered which requires tile</em>) – tile registration to the reference. Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.register_overview_to_reference">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">register_overview_to_reference</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channel</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overview_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DAPI_1_1'</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.register_overview_to_reference" title="Link to this definition">¶</a></dt>
<dd><p>Register the overview to the reference image</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Number of the ROI</p></li>
<li><p><strong>channel</strong> (<em>int</em>) – Channel to use for registration.</p></li>
<li><p><strong>downsample</strong> (<em>int</em><em>, </em><em>optional</em>) – Downsample factor. Defaults to 3.</p></li>
<li><p><strong>overview_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the overview image. Defaults to
“DAPI_1_1”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Shift in x and y
final_shape (tuple): Final shape of the stitched images
stitched_fixed (np.ndarray): Stitched reference image
stitched_moving (np.ndarray): Stitched overview image</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>shift (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.ara_registration.spots_ara_infos">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.ara_registration.</span></span><span class="sig-name descname"><span class="pre">spots_ara_infos</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spots</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atlas_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">acronyms</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inplace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">full_scale_coordinates</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.ara_registration.spots_ara_infos" title="Link to this definition">¶</a></dt>
<dd><p>Add ARA coordinates and area ID to spots dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>spots</strong> (<em>pd.DataFrame</em>) – Spots dataframe</p></li>
<li><p><strong>atlas_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Atlas size (10, 25 or 50) for find areas borders.
Defaults to 10</p></li>
<li><p><strong>acronyms</strong> (<em>bool</em><em>, </em><em>optional</em>) – Add an acronym column with area name. Defaults to
False.</p></li>
<li><p><strong>inplace</strong> (<em>bool</em><em>, </em><em>optional</em>) – add the column to spots inplace or return a copy.
Defaults to True</p></li>
<li><p><strong>full_scale_coordinates</strong> (<em>bool</em><em>, </em><em>optional</em>) – If true, use the full scale image to
find coordinates, otherwise the downsample version used for registration.
Defaults to False.</p></li>
<li><p><strong>reload</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, reload the area image, otherwise recompute it.
Valid only if full_scale is False. Defaults to True.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Print progress. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>reference or copy of spots dataframe with four more</dt><dd><p>columns: <cite>ara_x</cite>, <cite>ara_y</cite>, <cite>ara_z</cite>, and <cite>area_id</cite></p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>spots (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.hybridisation">
<span id="iss-preprocess-pipeline-hybridisation-module"></span><h2>iss_preprocess.pipeline.hybridisation module<a class="headerlink" href="#module-iss_preprocess.pipeline.hybridisation" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.estimate_channel_correction_hybridisation">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">estimate_channel_correction_hybridisation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.estimate_channel_correction_hybridisation" title="Link to this definition">¶</a></dt>
<dd><p>Compute grayscale value distribution and normalisation factors for
all hybridisation rounds.</p>
<p>Each <cite>correction_tiles</cite> of <cite>ops</cite> is filtered before being used to compute the
distribution of pixel values.
Normalisation factor to equalise these distribution across channels and rounds are
defined as <cite>ops[“correction_quantile”]</cite> of the distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em><em> or </em><em>Path</em>) – Relative path to the data folder</p></li>
<li><p><strong>prefix</strong> (<em>list</em><em>, </em><em>optional</em>) – List of prefix of hybridisation rounds to process. If
None, all hybridisation rounds are processed. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A 65536 x Nch x Nrounds distribution of grayscale values</dt><dd><p>for filtered stacks</p>
</dd>
</dl>
<p>norm_factors (np.array) A Nch x Nround array of normalisation factors</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pixel_dist (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.extract_hyb_spots_all">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">extract_hyb_spots_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.extract_hyb_spots_all" title="Link to this definition">¶</a></dt>
<dd><p>Start <cite>sbatch</cite> jobs to detect hybridisation spots for each hybridisation
round and ROI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.extract_hyb_spots_roi">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">extract_hyb_spots_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.extract_hyb_spots_roi" title="Link to this definition">¶</a></dt>
<dd><p>Detect hybridisation spots for a given hybridisation round and ROI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the hybridisation round, e.g. “hybridisation_1_1”.</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – ID of the ROI to process, as specified in MicroManager
(i.e. 1-based)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.extract_hyb_spots_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">extract_hyb_spots_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.extract_hyb_spots_tile" title="Link to this definition">¶</a></dt>
<dd><p>Detect hybridisation spots for a given tile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – Coordinates of tile to load: ROI, Xpos, Ypos.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the hybridisation round, e.g. “hybridisation_1_1”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.hyb_spot_cluster_means">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">hyb_spot_cluster_means</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.hyb_spot_cluster_means" title="Link to this definition">¶</a></dt>
<dd><p>Estimate bleedthrough matrices for hybridisation spots. Spot
colors for each dye are initialized based on the metadata in the
hybridisation probe list.</p>
<p>Uses tiles specified in <cite>ops[“barcode_ref_tiles”]</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of hybridisation round, e.g. “hybridisation_1_1”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Nprobes x Nch bleedthrough matrix.
pandas.DataFrame: DataFrame of all detected spots across all tiles.
list: list of gene names based on probe metadata.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.load_and_register_hyb_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">load_and_register_hyb_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1,</span> <span class="pre">0,</span> <span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hybridisation_1_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(2,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_channels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_shifts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'best'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.load_and_register_hyb_tile" title="Link to this definition">¶</a></dt>
<dd><p>Load hybridisation tile and align channels. Optionally, filter, correct
illumination and channel brightness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em><em>, </em><em>options</em>) – Coordinates of tile to load: ROI, Xpos, Ypos.
Defaults to (1, 0, 0).</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the hybridisation round.
Defaults to “hybridisation_1_1”.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Filename suffix corresponding to the z-projection
to use. Defaults to “fstack”.</p></li>
<li><p><strong>filter_r</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Inner and out radius for the hanning filter.
If <cite>False</cite>, stack is not filtered. Defaults to (2, 4).</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to correct vignetting.
Defaults to False.</p></li>
<li><p><strong>correct_channels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to normalize channel brightness.
Defaults to False.</p></li>
<li><p><strong>correct_shifts</strong> (<em>str</em><em>, </em><em>optional</em>) – Which shift to use. One of <cite>reference</cite>,
<cite>single_tile</cite>, <cite>ransac</cite>, or <cite>best</cite>. Defaults to ‘best’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>X x Y x Nch image stack.
numpy.ndarray: X x Y boolean mask, identifying bad pixels that we were not</p>
<blockquote>
<div><p>imaged for all channels (due to registration offsets) and should be
discarded during analysis.</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.hybridisation.setup_hyb_spot_calling">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.hybridisation.</span></span><span class="sig-name descname"><span class="pre">setup_hyb_spot_calling</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vis</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.hybridisation.setup_hyb_spot_calling" title="Link to this definition">¶</a></dt>
<dd><p>Prepare and save bleedthrough matrices for hybridisation rounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>list</em><em>, </em><em>optional</em>) – List of prefix of hybridisation rounds to process.
If None, all hybridisation rounds are processed. Defaults to None.</p></li>
<li><p><strong>vis</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to generate diagnostic plots. Defaults to True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.pipeline">
<span id="iss-preprocess-pipeline-pipeline-module"></span><h2>iss_preprocess.pipeline.pipeline module<a class="headerlink" href="#module-iss_preprocess.pipeline.pipeline" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.call_spots">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">call_spots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hybridisation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">setup_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.call_spots" title="Link to this definition">¶</a></dt>
<dd><p>Master method to run spot calling.</p>
<p>Must be run after <cite>iss project-and-average</cite> and <cite>iss register</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data folder</p></li>
<li><p><strong>genes</strong> (<em>bool</em><em>, </em><em>optional</em>) – Run genes spot calling. Defaults to True.</p></li>
<li><p><strong>barcodes</strong> (<em>bool</em><em>, </em><em>optional</em>) – Run barcode calling. Defaults to True.</p></li>
<li><p><strong>hybridisation</strong> (<em>bool</em><em>, </em><em>optional</em>) – Run hybridisation spot calling. Defaults to True</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo all processing steps? Defaults to False.</p></li>
<li><p><strong>setup_only</strong> (<em>bool</em><em>, </em><em>optional</em>) – Only setup the spot calling, do not run it.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use SLURM to run the jobs. Defaults to
True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.correct_shifts">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">correct_shifts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.correct_shifts" title="Link to this definition">¶</a></dt>
<dd><p>Correct X-Y shifts using robust regression across tiles.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.create_all_single_averages">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">create_all_single_averages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_batch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">todo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('genes_rounds',</span> <span class="pre">'barcode_rounds',</span> <span class="pre">'fluorescence',</span> <span class="pre">'hybridisation')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_average</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.create_all_single_averages" title="Link to this definition">¶</a></dt>
<dd><p>Average all tiffs in each folder and then all folders by acquisition type</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to data, relative to project.</p></li>
<li><p><strong>n_batch</strong> (<em>int</em>) – Number of batch to average before taking their median.
If None, will do as many batches as images.</p></li>
<li><p><strong>todo</strong> (<em>tuple</em>) – type of acquisition to process. Default to <cite>(“genes_rounds”,
“barcode_rounds”, “fluorescence”, “hybridisation”)</cite>. Ignored if <cite>to_average</cite>
is not None.</p></li>
<li><p><strong>to_average</strong> (<em>list</em><em>, </em><em>optional</em>) – List of folders to average. If None, will average
all folders listed in metadata. Defaults to None.</p></li>
<li><p><strong>dependency</strong> (<em>list</em><em>, </em><em>optional</em>) – List of job IDs to wait for before starting the
current job. Defaults to None.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Submit jobs to slurm. Defaults to True.</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo if the average already exists. Defaults to
False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.create_grand_averages">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">create_grand_averages</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix_todo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('genes_round',</span> <span class="pre">'barcode_round',</span> <span class="pre">'')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix_todo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('max',</span> <span class="pre">'median')</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_batch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.create_grand_averages" title="Link to this definition">¶</a></dt>
<dd><p>Average single acquisition averages into grand average</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the folder, relative to <cite>projects</cite> folder</p></li>
<li><p><strong>suffix</strong> (<em>str</em>) – Projection suffix to filter tifs. Defaults to None.</p></li>
<li><p><strong>prefix_todo</strong> (<em>tuple</em><em>, </em><em>optional</em>) – List of str, names of the tifs to average. An
empty string will average all tifs. Defaults to (“genes_round”,
“barcode_round”, “”).</p></li>
<li><p><strong>suffix_todo</strong> (<em>list</em><em>, </em><em>optional</em>) – List of str, suffixes to filter tifs. Defaults to
(‘max’, ‘median’).</p></li>
<li><p><strong>n_batch</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of batch to average before taking their median.
If None, will do as many batches as images. Defaults to None.</p></li>
<li><p><strong>dependency</strong> (<em>list</em><em>, </em><em>optional</em>) – List of job IDs to wait for before starting the
current job. Defaults to None.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Submit jobs to slurm. Defaults to True.</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo if the average already exists. Defaults to
False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.create_single_average">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">create_single_average</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subfolder</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">subtract_black</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_batch</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_fname</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">combine_tilestats</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">exclude_tiffs</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.create_single_average" title="Link to this definition">¶</a></dt>
<dd><p>Create normalised average of all tifs in a single folder.</p>
<p>If prefix_filter is not None, the output will be “{prefix_filter}_average.tif”,
otherwise it will be “{folder_path.name}_average.tif”</p>
<dl class="simple">
<dt>Other arguments are read from <cite>ops</cite>:</dt><dd><p>average_clip_value: Value to clip images before averaging.
normalise: Normalise output maximum to one.</p>
</dd>
</dl>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the acquisition folder, relative to <cite>projects</cite> folder</p></li>
<li><p><strong>subfolder</strong> (<em>str</em>) – subfolder in folder_path containing the tifs to average.</p></li>
<li><p><strong>subtract_black</strong> (<em>bool</em>) – Subtract black level (read from <cite>ops</cite>)</p></li>
<li><p><strong>n_batch</strong> (<em>int</em>) – Number of batch to average before taking their median.
If None, will do as many batches as images.</p></li>
<li><p><strong>prefix_filter</strong> (<em>str</em><em>, </em><em>optional</em>) – prefix name to filter tifs. Only file starting
with <cite>prefix</cite> will be averaged. Defaults to None.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – suffix to filter tifs. Defaults to None</p></li>
<li><p><strong>target_fname</strong> (<em>str</em><em>, </em><em>optional</em>) – Target file name to save the average. Defaults to
None</p></li>
<li><p><strong>combine_tilestats</strong> (<em>bool</em><em>, </em><em>optional</em>) – Compute new tilestats distribution of
averaged images if True, combine pre-existing tilestats into one otherwise.
Defaults to False</p></li>
<li><p><strong>exclude_tiffs</strong> (<em>list</em><em>, </em><em>optional</em>) – List of str filter to exclude tiffs from average</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Average image
np.array: Distribution of pixel values</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.overview_for_ara_registration">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">overview_for_ara_registration</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rois_to_do</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sigma_blur</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">non_similar_overview</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.overview_for_ara_registration" title="Link to this definition">¶</a></dt>
<dd><p>Generate a stitched overview for registering to the ARA</p>
<p>ABBA requires pyramidal OME-TIFF with resolution information. We will generate such
stitched files and save them with a log yaml file indicating info about downsampling</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data folder</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition to use for the overview e.g. <cite>genes_round_1_1</cite></p></li>
<li><p><strong>rois_to_do</strong> (<em>list</em><em>, </em><em>optional</em>) – ROIs to process. If None (default), process all
ROIs</p></li>
<li><p><strong>sigma_blur</strong> (<em>float</em><em>, </em><em>optional</em>) – sigma of the gaussian filter, in downsampled
pixel size. Defaults to 10</p></li>
<li><p><strong>ref_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the reference coordinates. Defaults to
<cite>genes_round</cite></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.project_and_average">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">project_and_average</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.project_and_average" title="Link to this definition">¶</a></dt>
<dd><p>Project and average all available data then create plots.</p>
<p>Creates a list of expected acquisition folders from metadata
Checks for the existence of expected folders in the raw data
and determines the completion status of each acquisition type.
Runs projection on unprojected data and reprojects failed tiles.
Creates averages of projections and then plots overview images.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo all processing steps? Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of job IDs for the slurm jobs created.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>po_job_ids (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.register_acquisition">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">register_acquisition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.register_acquisition" title="Link to this definition">¶</a></dt>
<dd><p>Register an acquisition across all rounds and channels</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>path</strong> (<em>str</em>) – Path to the data folder</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the acquisition to register</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo if files exist. Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.register_reference_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">register_reference_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">diag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.register_reference_tile" title="Link to this definition">¶</a></dt>
<dd><p>Register the reference tile across channels and rounds</p>
<p>This function estimates the shifts and rotations between rounds and
channels using the reference tile and generates diagnostic plots if
requested.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory prefix to use, e.g. ‘genes_round’.
Defaults to ‘genes_round’.</p></li>
<li><p><strong>diag</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save diagnostic plots. Defaults to False.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Submit job to slurm. Defaults to True.</p></li>
<li><p><strong>redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo if files exist. Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.segment_and_stitch_mcherry_cells">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">segment_and_stitch_mcherry_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.segment_and_stitch_mcherry_cells" title="Link to this definition">¶</a></dt>
<dd><p>Master function for mCherry cell segmentation and stitching</p>
<p>Will call in turn the following functions:
- <cite>segment_mcherry_cells</cite>
- <cite>filter_mcherry_cells</cite> if ops[‘filter_mask’] is True
- <cite>register_within</cite> to find overlapping region (with reload=True)
- <cite>remove_duplicate</cite>
- <cite>stitch_mcherry_cells</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data folder</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the mCherry acquisition</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use SLURM to run the jobs. Defaults to
True.</p></li>
<li><p><strong>slurm_folder</strong> (<em>str</em><em>, </em><em>optional</em>) – Folder to save SLURM logs. Defaults to None.</p></li>
<li><p><strong>job_dependency</strong> (<em>list</em><em>, </em><em>optional</em>) – List of job IDs to wait for before starting the</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.pipeline.setup_channel_correction">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.pipeline.</span></span><span class="sig-name descname"><span class="pre">setup_channel_correction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix_to_do</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.pipeline.setup_channel_correction" title="Link to this definition">¶</a></dt>
<dd><p>Setup channel correction for barcode, genes and hybridisation rounds</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data folder</p></li>
<li><p><strong>prefix_to_do</strong> (<em>list</em><em>, </em><em>optional</em>) – Prefixes to process. Defaults to None.</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Redo all processing steps? Defaults to False.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use SLURM to run the jobs. Defaults to
True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of job IDs for the slurm jobs created</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.project">
<span id="iss-preprocess-pipeline-project-module"></span><h2>iss_preprocess.pipeline.project module<a class="headerlink" href="#module-iss_preprocess.pipeline.project" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.check_projection">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">check_projection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffixes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">('max',</span> <span class="pre">'median')</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.check_projection" title="Link to this definition">¶</a></dt>
<dd><p>Check if all tiles have been projected successfully.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition prefix, e.g. “genes_round_1_1”.</p></li>
<li><p><strong>suffixes</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Projection suffixes to check for.</p></li>
<li><p><strong>to</strong> (<em>Defaults</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.check_roi_dims">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">check_roi_dims</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.check_roi_dims" title="Link to this definition">¶</a></dt>
<dd><p>Check if all ROI dimensions are the same across rounds.
:param data_path: Relative path to data.
:type data_path: str</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If ROI dimensions are not the same across rounds.</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.project_round">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">project_round</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.project_round" title="Link to this definition">¶</a></dt>
<dd><p>Start SLURM jobs to z-project all tiles from a single imaging round.
Also, copy one of the MicroManager metadata files from raw to processed directory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to dataset.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Full folder name prefix, including round number.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to re-project if files already exist.
Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.project_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">project_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">fname</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ops</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sth</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">13</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.project_tile" title="Link to this definition">¶</a></dt>
<dd><p>Calculates projections for a single tile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>fname</strong> (<em>str</em>) – path to tile <em>without</em> <cite>‘.ome.tif’</cite> extension.</p></li>
<li><p><strong>ops</strong> (<em>dict</em>) – dictionary of values from the ops file.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether to repeat if already completed. Defaults to
False.</p></li>
<li><p><strong>sth</strong> (<em>int</em><em>, </em><em>optional</em>) – size of the structuring element for the fstack projection.
Used only if <cite>make_fstack</cite> is True. Defaults to 13.</p></li>
<li><p><strong>target_name</strong> (<em>str</em><em>, </em><em>optional</em>) – name of the target file. If None, it will be the
same as the input file. Defaults to None.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – print progress. Defaults to True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.project_tile_by_coors">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">project_tile_by_coors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.project_tile_by_coors" title="Link to this definition">¶</a></dt>
<dd><p>Project a single tile by its coordinates.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – (roi, x, y) coordinates of the tile.</p></li>
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition prefix, e.g. “genes_round_1_1”.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to re-project if files already exist.
Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.project_tile_row">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">project_tile_row</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_row</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_col</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.project_tile_row" title="Link to this definition">¶</a></dt>
<dd><p>Calculate max intensity and extended DOF projections for a row of tiles in an ROI</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – relative path to dataset</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – directory / file name prefix, e.g. ‘gene_round’</p></li>
<li><p><strong>tile_roi</strong> (<em>int</em>) – index of the ROI</p></li>
<li><p><strong>tile_row</strong> (<em>int</em>) – index of the row to process</p></li>
<li><p><strong>max_col</strong> (<em>int</em>) – Maximum columns index. Column 0 to max_col will be projected.</p></li>
<li><p><strong>overwrite</strong> (<em>bool</em><em>, </em><em>optional</em>) – whether to redo projection if files already exist.
Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.project.reproject_failed">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.project.</span></span><span class="sig-name descname"><span class="pre">reproject_failed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.project.reproject_failed" title="Link to this definition">¶</a></dt>
<dd><p>Re-project tiles that failed to project previously.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.register">
<span id="iss-preprocess-pipeline-register-module"></span><h2>iss_preprocess.pipeline.register module<a class="headerlink" href="#module-iss_preprocess.pipeline.register" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.correct_hyb_shifts">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">correct_hyb_shifts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.correct_hyb_shifts" title="Link to this definition">¶</a></dt>
<dd><p>Use robust regression across tiles to correct shifts and angles
for hybridisation rounds. Either processes a specific hybridisation
round or all rounds.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Directory prefix to use, e.g. “hybridisation_1_1”. If None,
processes all hybridisation acquisitions.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.correct_shifts_roi">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">correct_shifts_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_dims</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_tiles</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.correct_shifts_roi" title="Link to this definition">¶</a></dt>
<dd><p>Use robust regression to correct shifts across tiles for a single ROI.</p>
<p>RANSAC regression is applied to shifts within and across channels using
tile X and Y position as predictors. This will load the <cite>single_tile</cite> shifts and
create the <cite>corrected</cite> shifts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>roi_dims</strong> (<em>tuple</em>) – Dimensions of the ROI to be processed, in (ROI_ID, Xtiles,
Ytiles) format.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory prefix to use. Defaults to “genes_round”.</p></li>
<li><p><strong>max_shift</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum shift to include tiles in RANSAC regression.
Tiles with larger absolute shifts will not be included in the fit but will
still have their corrected shifts estimated. Defaults to 500.</p></li>
<li><p><strong>min_tiles</strong> (<em>int</em><em>, </em><em>optional</em>) – Minimum number of tiles to use for RANSAC regression,
otherwise median is used.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.correct_shifts_single_round_roi">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">correct_shifts_single_round_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_dims</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'hybridisation_1_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">500</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_angle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">align_method</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_chans</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.correct_shifts_single_round_roi" title="Link to this definition">¶</a></dt>
<dd><p>Use robust regression across tiles to correct shifts and angles
for a single hybridisation round and ROI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>roi_dims</strong> (<em>tuple</em>) – Dimensions of the ROI to be processed, in (ROI_ID, Xtiles,
Ytiles) format.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the round to be processed.
Defaults to “hybridisation_1_1”.</p></li>
<li><p><strong>max_shift</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum shift to include tiles in RANSAC regression.
Tiles with larger absolute shifts will not be included in the fit but will
still have their corrected shifts estimated. Defaults to 500.</p></li>
<li><p><strong>fit_angle</strong> (<em>bool</em><em>, </em><em>optional</em>) – Fit the angle with robust regression if True,
otherwise takes the median. Defaults to True</p></li>
<li><p><strong>align_method</strong> (<em>str</em><em>, </em><em>optional</em>) – Method to use for alignment. If None, will be
read from ops. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.correct_shifts_to_ref">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">correct_shifts_to_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_angle</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.correct_shifts_to_ref" title="Link to this definition">¶</a></dt>
<dd><p>Use robust regression across tiles to correct shifts to reference acquisition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Directory prefix to use, e.g. “genes_round”.</p></li>
<li><p><strong>fit_angle</strong> (<em>bool</em><em>, </em><em>optional</em>) – Fit the angle with robust regression if True,
otherwise takes the median. Defaults to False</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.estimate_shifts_by_coors">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">estimate_shifts_by_coors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0,</span> <span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.estimate_shifts_by_coors" title="Link to this definition">¶</a></dt>
<dd><p>Estimate shifts across channels and sequencing rounds using provided reference
rotation angles and scale factors.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Coordinates of tile to register, in (ROI, X, Y)
format. Defaults to (0, 0, 0).</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory prefix to register. Defaults to “genes_round”.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Filename suffix specifying which z-projection to use.
Defaults to “fstack”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.filter_ransac_shifts">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">filter_ransac_shifts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi_dims</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_residuals</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.filter_ransac_shifts" title="Link to this definition">¶</a></dt>
<dd><p>Filter shifts to use RANSAC shifts only if the initial shifts are off</p>
<p>This will load the <cite>single_tile</cite> and <cite>corrected</cite> shifts and create the <cite>best</cite> shifts</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Directory prefix to use, e.g. “genes_round”.</p></li>
<li><p><strong>roi_dims</strong> (<em>tuple</em>) – Dimensions of the ROI to be processed, in (ROI_ID, Xtiles,
Ytiles)</p></li>
<li><p><strong>max_residuals</strong> (<em>int</em><em>, </em><em>optional</em>) – Threshold on residuals above which the RANSAC
shifts are used. Defaults to 10.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.load_and_register_raw_stack">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">load_and_register_raw_stack</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_shifts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.load_and_register_raw_stack" title="Link to this definition">¶</a></dt>
<dd><p>Load a raw stack and apply channel registration.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition to load.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – (Roi, tileX, tileY) tuple</p></li>
<li><p><strong>corrected_shifts</strong> (<em>str</em><em>, </em><em>optional</em>) – Shift correction method. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A (X x Y x Nchannels) registered stack</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.load_and_register_sequencing_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">load_and_register_sequencing_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(1,</span> <span class="pre">0,</span> <span class="pre">0)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(2,</span> <span class="pre">4)</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_channels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_shifts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'best'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrounds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">specific_rounds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.load_and_register_sequencing_tile" title="Link to this definition">¶</a></dt>
<dd><p>Load sequencing tile and align channels. Optionally, filter, correct
illumination and channel brightness.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em><em>, </em><em>options</em>) – Coordinates of tile to load: ROI, Xpos, Ypos.
Defaults to (1, 0, 0).</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the sequencing round.
Defaults to “genes_round”.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Filename suffix corresponding to the z-projection
to use. Defaults to “fstack”.</p></li>
<li><p><strong>filter_r</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Inner and out radius for the hanning filter.
If <cite>False</cite>, stack is not filtered. Defaults to (2, 4).</p></li>
<li><p><strong>correct_channels</strong> (<em>bool</em><em> or </em><em>str</em><em>, </em><em>optional</em>) – Whether to normalize channel
brightness. If ‘round1_only’, normalise by round 1 correction factor,
otherwise, if True use all norm_factors. Defaults to False.</p></li>
<li><p><strong>corrected_shifts</strong> (<em>str</em><em>, </em><em>optional</em>) – Which shift to use. One of <cite>reference</cite>,
<cite>single_tile</cite>, <cite>ransac</cite>, or <cite>best</cite>. Defaults to ‘best’.</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to correct vignetting.
Defaults to False.</p></li>
<li><p><strong>nrounds</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of sequencing rounds to load. Used only if
specific_rounds is None. Defaults to 7.</p></li>
<li><p><strong>specific_rounds</strong> (<em>list</em><em>, </em><em>optional</em>) – if not None, specifies which rounds must be
loaded and ignores <cite>nrounds</cite>. Defaults to None</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>X x Y x Nch x len(specific_rounds) or Nrounds image stack.
numpy.ndarray: X x Y boolean mask, identifying bad pixels that we were not</p>
<blockquote>
<div><p>imaged for all channels and rounds (due to registration offsets) and should
be discarded during analysis.</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.load_and_register_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">load_and_register_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">zero_bad_pixels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.load_and_register_tile" title="Link to this definition">¶</a></dt>
<dd><p>Load one single tile</p>
<p>Load a tile of <cite>prefix</cite> with channels/rounds registered, apply illumination
correction and filtering.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – (Roi, tileX, tileY) tuple</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition to load. If <cite>genes_round</cite> or <cite>barcode_round</cite> will load
all the rounds.</p></li>
<li><p><strong>filter_r</strong> (<em>bool</em><em>, </em><em>optional</em>) – Apply filter on rounds data? Parameters will be read
from <cite>ops</cite>. Default to True</p></li>
<li><p><strong>projection</strong> (<em>str</em><em>, </em><em>optional</em>) – Projection to use. If None, will read from <cite>ops</cite>.
Defaults to None</p></li>
<li><p><strong>zero_bad_pixels</strong> (<em>bool</em><em>, </em><em>optional</em>) – Set bad pixels to zero. Defaults to False</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Apply illumination correction. Defaults
to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A (X x Y x Nchannels x Nrounds) registered stack
numpy.ndarray: X x Y boolean mask of bad pixels where data is missing after</p>
<blockquote>
<div><p>registration</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.merge_shifts">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">merge_shifts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_chans</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.merge_shifts" title="Link to this definition">¶</a></dt>
<dd><p>Merge shifts for all ROI/tiles into a single shift median shift</p>
<p>Useful if some of the registration failed and we want to use the same shift for all
tiles</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Directory prefix to use, e.g. “hybridisation_1_1”.</p></li>
<li><p><strong>n_chans</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of channels to merge. Defaults to 4.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.register_channels_by_pairs">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">register_channels_by_pairs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">channel_grouping</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ops</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ops_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">stack</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binarise_quantile</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_tforms</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debug</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.register_channels_by_pairs" title="Link to this definition">¶</a></dt>
<dd><p>Register channels for a single tile iteratively by group of channels</p>
<p>channel_grouping must be a list of list (of list ….). The inner most levels will
be registered together, using the first channel of the list as reference. Then the
upper level will be registered together.
For instance, if channel_grouping = [[0, 1], [2, 3]], channels 0 and 1 will be
registered together (ref=0), then channels 2 and 3 will be registered together
(ref=2), and finally the two groups will be registered together (ref=0).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>channel_grouping</strong> (<em>list</em>) – List of list of channels to register together.</p></li>
<li><p><strong>ops</strong> (<em>dict</em>) – Experiment metadata.</p></li>
<li><p><strong>ops_prefix</strong> (<em>str</em>) – Prefix to use for ops, e.g. “genes”.</p></li>
<li><p><strong>stack</strong> (<em>np.array</em>) – Image stack to register.</p></li>
<li><p><strong>reference_prefix</strong> (<em>str</em>) – Prefix to load scale or initial matrix from.</p></li>
<li><p><strong>binarise_quantile</strong> (<em>float</em>) – Quantile to binarise images before registration.</p></li>
<li><p><strong>reference_tforms</strong> (<em>dict</em>) – Reference transformation parameters.</p></li>
<li><p><strong>debug</strong> (<em>bool</em>) – Return debug information.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Transformation parameters.
dict: Debug information, only if debug is True</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.register_fluorescent_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">register_fluorescent_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debug</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.register_fluorescent_tile" title="Link to this definition">¶</a></dt>
<dd><p>Estimate channel registration parameters for a single round acquisition</p>
<p>The stack will be binarised if ops[f”{prefix_start}_binarise_quantile”] is not None.
The scale and initial parameters will be loaded from the reference prefix and
optimised using either a similarity transform or an affine transform, depending
on ops[“align_method”].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – Coordinates of tile to register, in (ROI, X, Y) format.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Directory prefix to register. Defaults to</p></li>
<li><p><strong>reference_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to load scale or initial matrix from.
Defaults to None.</p></li>
<li><p><strong>debug</strong> (<em>bool</em><em>, </em><em>optional</em>) – Return debug information. Defaults to False.</p></li>
<li><p><strong>save_output</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save output to disk. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Debug information if debug is True, None otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.run_correct_shifts">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">run_correct_shifts</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.run_correct_shifts" title="Link to this definition">¶</a></dt>
<dd><p>Use robust regression to correct shifts across tiles within an ROI
for all ROIs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Directory prefix to use, e.g. “genes_round”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.register.run_register_reference_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.register.</span></span><span class="sig-name descname"><span class="pre">run_register_reference_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">diag</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.register.run_register_reference_tile" title="Link to this definition">¶</a></dt>
<dd><p>Subfunction to run the registration of the reference tile</p>
<p>This function actually perform the computation. It performs the registration of the
the reference tile specified inn the ops. This include shifts and rotations between
rounds and shifts, rotations, and scaling between channels.</p>
<p>Shifts are estimated using phase correlation. Rotation and scaling are estimated
using iterative grid search.</p>
<p>Results are saved in a npz file in the processed directory in:
data_path / ‘reg’ / <cite>prefix</cite> / ‘ref_tile_tforms_`prefix`_round.npz’</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Directory prefix to register.
Defaults to “genes_round”.</p></li>
<li><p><strong>diag</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save diagnostic plots.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.segment">
<span id="iss-preprocess-pipeline-segment-module"></span><h2>iss_preprocess.pipeline.segment module<a class="headerlink" href="#module-iss_preprocess.pipeline.segment" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.add_mask_id">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">add_mask_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_dot_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.15</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spot_score_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hyb_score_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_hyb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_barcodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.add_mask_id" title="Link to this definition">¶</a></dt>
<dd><p>Load gene, barcode, and hybridisation spots and add a mask_id column to each
spots dataframe</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – ID of the ROI to load</p></li>
<li><p><strong>masks</strong> (<em>np.array</em>) – Array of labels.</p></li>
<li><p><strong>barcode_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Rabies barcode dataframe, if None, will
load “barcode_df_roi{roi}.pkl”. Defaults to None.</p></li>
<li><p><strong>barcode_dot_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for the barcode dot product.
Only spots above the threshold will be counted. Defaults to 0.15.</p></li>
<li><p><strong>spot_score_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for the OMP score. Only spots
above the threshold will be counted. Defaults to 0.1.</p></li>
<li><p><strong>hyb_score_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for hybridisation spots. Only
spots above the threshold will be counted. Defaults to 0.8.</p></li>
<li><p><strong>load_genes</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to load gene spots. Defaults to True.</p></li>
<li><p><strong>load_hyb</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to load hybridisation spots. Defaults to True</p></li>
<li><p><strong>load_barcodes</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to load barcode spots. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary of spots dataframes</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.filter_mcherry_cells">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">filter_mcherry_cells</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.filter_mcherry_cells" title="Link to this definition">¶</a></dt>
<dd><p>Use GMM to cluster cells and remove non-cell masks.</p>
<p>This function will:
- Use all saved dataframe to fit a GMM model, calling _gmm_cluster_mcherry_cells
- Apply this model to all masks, calling _remove_non_cell_masks</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
<li><p><strong>tile_list</strong> (<em>list</em><em>, </em><em>optional</em>) – List of tiles to process. If None, will process all
tiles. Defaults to None.</p></li>
<li><p><strong>use_rois</strong> (<em>list</em><em>, </em><em>optional</em>) – List of ROIs to process. If None, will process all
ROIs. Used only if tile_list is None. Defaults to None.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use slurm to parallelize the process.
Defaults to True.</p></li>
<li><p><strong>slurm_folder</strong> (<em>str</em><em>, </em><em>optional</em>) – Folder to save slurm logs. Defaults to None.</p></li>
<li><p><strong>job_dependency</strong> (<em>list</em><em>, </em><em>optional</em>) – List of job ids to wait for before starting
the job. Defaults to None.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.find_edge_touching_masks">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">find_edge_touching_masks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">masks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">border_width</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">4</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.find_edge_touching_masks" title="Link to this definition">¶</a></dt>
<dd><p>Finds masks that touch the edge of the image.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>masks</strong> (<em>np.ndarray</em>) – The binary or labeled mask array where each cell is
represented by a unique integer, and background is 0.</p></li>
<li><p><strong>border_width</strong> (<em>int</em>) – The width of the border to consider when checking for edge
touching. Defaults to 4.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A list of unique labels that touch the edge of the
image.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>edge_touching_labels (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.get_big_masks">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">get_big_masks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_expansion</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.get_big_masks" title="Link to this definition">¶</a></dt>
<dd><p>Small internal function to avoid code duplication</p>
<p>Reload and expand masks if needed</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>masks</strong> (<em>np.array</em>) – Array of labels.</p></li>
<li><p><strong>mask_expansion</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance in um to expand masks before counting
rolonies per cells. None for no expansion. Defaults to 5.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>masks expanded</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.get_cell_masks">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">get_cell_masks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'corrected'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_expansion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">curated</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.get_cell_masks" title="Link to this definition">¶</a></dt>
<dd><p>Small wrapper to get cell masks from a given data path.</p>
<p>Wrap to ensure we use the same projection for all calls</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to acquisition data (chamber folder)</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Region of interest</p></li>
<li><p><strong>projection</strong> (<em>str</em><em>, </em><em>optional</em>) – Projection to use. Defaults to “corrected”.</p></li>
<li><p><strong>mask_expansion</strong> (<em>int</em><em>, </em><em>optional</em>) – Expansion of the mask. If None, reads from ops.
Defaults to None.</p></li>
<li><p><strong>reload</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, reload the saved masks, otherwise regenerate
from individual tiles. Defaults to True.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to use for the masks. If None, reads from ops.
Defaults to None.</p></li>
<li><p><strong>curated</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use curated masks. These are manually
curated and have the same filename with “_curated.tif”. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Cell masks</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.get_overlap_regions">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">get_overlap_regions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_coors</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.get_overlap_regions" title="Link to this definition">¶</a></dt>
<dd><p>Determine the coordinates of the overlap region between two adjacent tiles using
explicit tile direction.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shifts</strong> (<em>dict</em>) – The dictionary containing the shift values for the down and right
tiles.</p></li>
<li><p><strong>tile_ref</strong> (<em>np.ndarray</em>) – The reference tile.</p></li>
<li><p><strong>tile_right</strong> (<em>np.ndarray</em>) – The right tile.</p></li>
<li><p><strong>tile_down</strong> (<em>np.ndarray</em>) – The down tile.</p></li>
<li><p><strong>tile_down_right</strong> (<em>np.ndarray</em>) – The down right tile.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>The overlap region between the reference tile and</dt><dd><p>the down tile.</p>
</dd>
<dt>overlap_down (np.ndarray): The overlap region between the down tile and the</dt><dd><p>reference tile.</p>
</dd>
<dt>overlap_ref_side (np.ndarray): The overlap region between the reference tile and</dt><dd><p>the right tile.</p>
</dd>
<dt>overlap_right (np.ndarray): The overlap region between the right tile and the</dt><dd><p>reference tile.</p>
</dd>
<dt>overlap_ref_with_down_right (np.ndarray): The overlap region between the</dt><dd><p>reference tile and the down right tile.</p>
</dd>
<dt>overlap_down_right_with_ref (np.ndarray): The overlap region between the down</dt><dd><p>right tile and the reference tile.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>overlap_ref_vert (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.get_stack_for_cellpose">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">get_stack_for_cellpose</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw_stack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.get_stack_for_cellpose" title="Link to this definition">¶</a></dt>
<dd><p>Load the stack to segment with cellpose.</p>
<p>This will load a stack with 2 channels from the raw data or the registered stack.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition prefix to use for segmentation.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – Coordinates of the tile to segment.</p></li>
<li><p><strong>use_raw_stack</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use the raw stack or the projected
stack. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>X x Y x channels (x Z) stack.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.make_cell_dataframe">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">make_cell_dataframe</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_expansion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">atlas_size</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">10</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.make_cell_dataframe" title="Link to this definition">¶</a></dt>
<dd><p>Make cell dataframe</p>
<p>The index will be the mask ID.
The dataframe will include, for each cell, their centroid, bounding box, and area.
If atlas_size is not None, it will also include the ID and acronym of the atlas
area where their centroid is located.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Number of the ROI to process</p></li>
<li><p><strong>masks</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Array of labels, if None will load <cite>masks_{roi}.npy</cite>
from the reg folder. Defaults to None.</p></li>
<li><p><strong>mask_expansion</strong> (<em>float</em><em>, </em><em>optional</em>) – Distance in um to expand masks before counting
rolonies per cells. None for no expansion. Defaults to None.</p></li>
<li><p><strong>atlas_size</strong> (<em>int</em><em>, </em><em>optional</em>) – Size of the atlas to use to load ARA information.
If None, will not get area information. Defaults to 10.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe with the cell information</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cell_df (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.remove_all_duplicate_masks">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">remove_all_duplicate_masks</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">upper_overlap_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.remove_all_duplicate_masks" title="Link to this definition">¶</a></dt>
<dd><p>Remove masks that overlap in adjacent tiles.</p>
<p>The <cite>within_acquisition</cite> registration must be run for <cite>prefix</cite> beforehand.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
<li><p><strong>upper_overlap_thresh</strong> (<em>float</em><em>, </em><em>optional</em>) – The upper threshold percentage for
considering mask overlap significant. If None, will use ops if defined,
0.3 otherwise. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A list of tuples containing the labels that</dt><dd><p>overlapped and their respective percentages.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>all_overlapping_pairs (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.run_cellpose_segmentation">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">run_cellpose_segmentation</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw_stack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.run_cellpose_segmentation" title="Link to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.run_mask_projection">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">run_mask_projection</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tx</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ty</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.run_mask_projection" title="Link to this definition">¶</a></dt>
<dd><p>Project masks to a single plane.</p>
<p>Wrapper around iss_preprocess.segment.cell.project_mask to run on slurm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition prefix to use for segmentation.</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – ROI ID to segment as specified in MicroManager (i.e. 1-based).</p></li>
<li><p><strong>tx</strong> (<em>int</em>) – X coordinate of the tile.</p></li>
<li><p><strong>ty</strong> (<em>int</em>) – Y coordinate of the tile.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>X x Y x channels (x Z) stack.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.save_curated_dataframes">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">save_curated_dataframes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">intensity_channels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mask_expansion</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.save_curated_dataframes" title="Link to this definition">¶</a></dt>
<dd><p>Save the curated dataframes to the cells folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
<li><p><strong>roi</strong> (<em>list</em><em>, </em><em>optional</em>) – List of ROIs to process. If None, will process all
ROIs. Defaults to None.</p></li>
<li><p><strong>mask_expansion</strong> (<em>int</em><em>, </em><em>optional</em>) – Mask expansion to use. If None, will use the
value from the ops. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe with the cell information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.save_mcherry_mask_df">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">save_mcherry_mask_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.save_mcherry_mask_df" title="Link to this definition">¶</a></dt>
<dd><p>Collate individual tile dataframes and remove overlapping masks.</p>
<p>This does not “stitch” the mask and keeps only the within tile x/y
coordinates, but it does precompute the stitched label.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to the data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe with the cell information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pd.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.save_unmixing_coefficients">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">save_unmixing_coefficients</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">signal_channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">seed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n_random</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.save_unmixing_coefficients" title="Link to this definition">¶</a></dt>
<dd><p>Find the unmixing coefficients.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the data directory.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
<li><p><strong>tile_coors</strong> (<em>list</em><em>, </em><em>optional</em>) – List of tile coordinates. If None, will use random
tiles. Defaults to None.</p></li>
<li><p><strong>background_channel</strong> (<em>int</em><em>, </em><em>optional</em>) – Channel index of the background image. If
None, will use the value from the ops. Defaults to None.</p></li>
<li><p><strong>signal_channel</strong> (<em>int</em><em>, </em><em>optional</em>) – Channel index of the signal image. If None, will
use the value from the ops. Defaults to None.</p></li>
<li><p><strong>projection</strong> (<em>str</em><em>, </em><em>optional</em>) – Projection method. If None, will use the value from
the ops. Defaults to None.</p></li>
<li><p><strong>seed</strong> (<em>int</em><em>, </em><em>optional</em>) – Random seed for the random tiles. If None, will use the
the value from the ops. Defaults to None.</p></li>
<li><p><strong>n_random</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of random tiles to use. If None, will use the
the value from the ops. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pure signal image.
coef (float): Unmixing coefficient.
intercept (float): Unmixing intercept</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pure_signal (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.segment_all_rois">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">segment_all_rois</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DAPI_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.segment_all_rois" title="Link to this definition">¶</a></dt>
<dd><p>Start batch jobs for segmentation for each ROI.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – acquisition prefix to use for segmentation.
Defaults to “DAPI_1”.</p></li>
<li><p><strong>use_gpu</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use GPU. Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.segment_all_tiles">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">segment_all_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DAPI_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_raw_stack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_rois</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rerun_cellpose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.segment_all_tiles" title="Link to this definition">¶</a></dt>
<dd><p>Start batch jobs for segmentation for each tile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – acquisition prefix to use for segmentation.
Defaults to “DAPI_1”.</p></li>
<li><p><strong>use_raw_stack</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use the raw stack and do 3d
segmentation. Defaults to True.</p></li>
<li><p><strong>use_gpu</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use GPU. Defaults to True.</p></li>
<li><p><strong>use_rois</strong> (<em>list</em><em>, </em><em>optional</em>) – List of ROIs to process. If None, will use all ROIs.
Defaults to None.</p></li>
<li><p><strong>tile_list</strong> (<em>list</em><em>, </em><em>optional</em>) – List of tiles to process. If provided will ignore
use_rois. If None, will use all tiles.</p></li>
<li><p><strong>rerun_cellpose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to rerun cellpose even if the raw masks
already exist (used only if use_raw_stack is True). Defaults to False.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use slurm. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of job IDs for the slurm jobs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.segment_mcherry_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">segment_mcherry_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tilex</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tiley</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.segment_mcherry_tile" title="Link to this definition">¶</a></dt>
<dd><p>Segment the mCherry channel of an image stack.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the data directory.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Region of interest.</p></li>
<li><p><strong>tilex</strong> (<em>int</em>) – X coordinate of the tile.</p></li>
<li><p><strong>tiley</strong> (<em>int</em>) – Y coordinate of the tile.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Binary image of the filtered masks.
filtered_df (pd.DataFrame): DataFrame of the filtered masks.
rejected_masks (np.ndarray): Binary image of the rejected masks.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>filtered_masks (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.segment_roi">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">segment_roi</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">iroi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'DAPI_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_gpu</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.segment_roi" title="Link to this definition">¶</a></dt>
<dd><p>Detect cells in a single ROI using Cellpose.</p>
<p>Much faster with GPU but requires very amount of VRAM for large ROIs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>iroi</strong> (<em>int</em>) – ROI ID to segment as specified in MicroManager (i.e. 1-based).</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Acquisition prefix to use for segmentation. Defaults to
“DAPI_1”.</p></li>
<li><p><strong>use_gpu</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to use GPU. Defaults to False.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.segment_spots">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">segment_spots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">masks</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_df</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">barcode_dot_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spot_score_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">hyb_score_threshold</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.8</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_genes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_hyb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">load_barcodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.segment_spots" title="Link to this definition">¶</a></dt>
<dd><p>Count number of rolonies per cell for barcodes and genes.</p>
<p>Only rolonies above the relevant threshold will be counted. (Note that genes
rolonies are already thresholded once after OMP).</p>
<p>Hybridisation and sequencing datasets will be fused.</p>
<p>Outputs are saved in the <cite>cells</cite> folder as f”genes_df_roi{roi}.pkl” and
f”barcode_df_roi{roi}.pkl”</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – ID of the ROI to load</p></li>
<li><p><strong>masks</strong> (<em>np.array</em><em>, </em><em>optional</em>) – Array of labels. If None will load using
“get_cell_masks”. Defaults to None.</p></li>
<li><p><strong>barcode_df</strong> (<em>pd.DataFrame</em><em>, </em><em>optional</em>) – Rabies barcode dataframe, if None, will
load “barcode_df_roi{roi}.pkl”. Defaults to None.</p></li>
<li><p><strong>barcode_dot_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for the barcode dot product.
Only spots above the threshold will be counted. Defaults to 0.15.</p></li>
<li><p><strong>spot_score_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for the OMP score. Only spots
above the threshold will be counted. Defaults to 0.1.</p></li>
<li><p><strong>hyb_score_threshold</strong> (<em>float</em><em>, </em><em>optional</em>) – Threshold for hybridisation spots. Only
spots above the threshold will be counted. Defaults to 0.8.</p></li>
<li><p><strong>load_genes</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to load gene spots. Defaults to True.</p></li>
<li><p><strong>load_hyb</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to load hybridisation spots. Defaults to True</p></li>
<li><p><strong>load_barcodes</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to load barcode spots. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Count of rolonies per barcode sequence per cell.</dt><dd><p>Index is the mask ID of the cell</p>
</dd>
<dt>fused_df (pd.DataFrame): Count of rolonies per genes or hybridisation probe per</dt><dd><p>cell. Index is the mask ID of the cell</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>barcode_df (pd.DataFrame)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.segment.unmix_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.segment.</span></span><span class="sig-name descname"><span class="pre">unmix_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">signal_channel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.segment.unmix_tile" title="Link to this definition">¶</a></dt>
<dd><p>Unmix one tile using the previously found coefficients.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Path to the data directory.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of the image stack.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – Coordinates of the tile.</p></li>
<li><p><strong>background_channel</strong> (<em>int</em><em>, </em><em>optional</em>) – Channel index of the background image.</p></li>
<li><p><strong>signal_channel</strong> (<em>int</em><em>, </em><em>optional</em>) – Channel index of the signal image.</p></li>
<li><p><strong>projection</strong> (<em>str</em><em>, </em><em>optional</em>) – Projection method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Unmixed image.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>unmixed (np.ndarray)</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.sequencing">
<span id="iss-preprocess-pipeline-sequencing-module"></span><h2>iss_preprocess.pipeline.sequencing module<a class="headerlink" href="#module-iss_preprocess.pipeline.sequencing" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.basecall_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">basecall_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_spots</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.basecall_tile" title="Link to this definition">¶</a></dt>
<dd><p>Detect and basecall barcodes for a given tile.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em><em>, </em><em>optional</em>) – Coordinates of tile to load: ROI, Xpos, Ypos.</p></li>
<li><p><strong>save_spots</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the detected spots. Used to run
without erasing during diagnostics. Defaults to True.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.compute_spot_sign_image">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">compute_spot_sign_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.compute_spot_sign_image" title="Link to this definition">¶</a></dt>
<dd><p>Compute the reference spot sign image to use in spot calling. Save it to
the processed data folder.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the sequencing read to use.
Defaults to “genes_round”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.detect_genes_on_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">detect_genes_on_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_stack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.detect_genes_on_tile" title="Link to this definition">¶</a></dt>
<dd><p>Apply the OMP algorithm to unmix spots in a given tile using the saved
gene dictionary and settings saved in <cite>ops.yml</cite>. Then detect gene spots in
the resulting gene maps.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – Coordinates of tile to load: ROI, Xpos, Ypos.</p></li>
<li><p><strong>save_stack</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save registered and preprocessed images.
Defaults to False.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the sequencing read to analyse.
Defaults to “genes_round”.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.estimate_channel_correction">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">estimate_channel_correction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">nrounds</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">7</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">fit_norm_factors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.estimate_channel_correction" title="Link to this definition">¶</a></dt>
<dd><p>Compute grayscale value distribution and normalisation factors</p>
<p>Each <cite>correction_tiles</cite> of <cite>ops</cite> is filtered before being used to compute the
distribution of pixel values.
Normalisation factor to equalise these distribution across channels and rounds are
defined as <cite>ops[“correction_quantile”]</cite> of the distribution.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em><em> or </em><em>Path</em>) – Relative path to the data folder</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Folder name prefix, before round number. Defaults
to “genes_round”.</p></li>
<li><p><strong>nrounds</strong> (<em>int</em><em>, </em><em>optional</em>) – Number of rounds. Defaults to 7.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A 65536 x Nch x Nrounds distribution of grayscale values</dt><dd><p>for filtered stacks</p>
</dd>
</dl>
<p>norm_factors (np.array) A Nch x Nround array of normalisation factors</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pixel_dist (np.array)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.get_reference_spots">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">get_reference_spots</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.get_reference_spots" title="Link to this definition">¶</a></dt>
<dd><p>Load the reference spots for the given dataset.</p>
<p>Internal function for setup_omp and setup_barcode_calling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Short prefix, either ‘genes’ or ‘barcode’. Defaults to
‘genes’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Detected spots.
list: Normalisation shifts.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.load_spot_sign_image">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">load_spot_sign_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">threshold</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_raw_image</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.load_spot_sign_image" title="Link to this definition">¶</a></dt>
<dd><p>Load the reference spot sign image to use in spot calling. First, check
if the spot sign image has been computed for the current dataset and use it
if available. Otherwise, use the spot sign image saved in the repo.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>threshold</strong> (<em>float</em>) – Absolute value threshold used to binarize the spot
sign image.</p></li>
<li><p><strong>return_raw_image</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to return the raw spot sign
image. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Spot sign image after thresholding, containing -1, 0, or 1s.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.run_omp_on_tile">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">run_omp_on_tile</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ops</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_stack</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.run_omp_on_tile" title="Link to this definition">¶</a></dt>
<dd><p>Run OMP on a tile and return the results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – Coordinates of the tile to process.</p></li>
<li><p><strong>ops</strong> (<em>dict</em>) – Dictionary of parameters.</p></li>
<li><p><strong>save_stack</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to save the registered stack.
Defaults to False.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of the sequencing read to use.
Defaults to “genes_round”.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>OMP results.
dict: Dictionary of OMP parameters.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.setup_barcode_calling">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">setup_barcode_calling</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.setup_barcode_calling" title="Link to this definition">¶</a></dt>
<dd><p>Detect spots and compute cluster means</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A list with Nrounds elements. Each a Nch x Ncl (square</dt><dd><p>because N channels is equal to N clusters) array of cluster means,
normalised by round 0 intensity</p>
</dd>
</dl>
<p>all_spots (pandas.DataFrame): All detected spots.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>cluster_means (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.sequencing.setup_omp">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.sequencing.</span></span><span class="sig-name descname"><span class="pre">setup_omp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">force_redo</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.sequencing.setup_omp" title="Link to this definition">¶</a></dt>
<dd><p>Prepare variables required to run the OMP algorithm. Finds isolated spots using
STD across rounds and channels. Detected spots are then used to determine the
bleedthrough matrix using scaled k-means.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>force_redo</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to redo the setup. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>N x M dictionary, where N = R * C and M is the</dt><dd><p>number of genes.</p>
</dd>
</dl>
<p>list: gene names.
float: norm shift for the OMP algorithm, estimated as median norm of all pixels.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline.stitch">
<span id="iss-preprocess-pipeline-stitch-module"></span><h2>iss_preprocess.pipeline.stitch module<a class="headerlink" href="#module-iss_preprocess.pipeline.stitch" title="Link to this heading">¶</a></h2>
<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.calculate_tile_positions">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">calculate_tile_positions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">shift_right</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shift_down</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_shape</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ntiles</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">x_direction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">y_direction</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.calculate_tile_positions" title="Link to this definition">¶</a></dt>
<dd><p>Calculate position of each tile based on the provided shifts.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>shift_right</strong> (<em>numpy.array</em>) – X and Y shifts between different columns. Either
a 2-element array or a ntiles[0] x ntiles[1] x 2 matrix of shifts</p></li>
<li><p><strong>shift_down</strong> (<em>numpy.array</em>) – X and Y shifts between different rows. Either
a 2-element array or a ntiles[0] x ntiles[1] x 2 matrix of shifts</p></li>
<li><p><strong>tile_shape</strong> (<em>numpy.array</em>) – shape of each tile</p></li>
<li><p><strong>ntiles</strong> (<em>numpy.array</em>) – number of tile rows and columns</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt><cite>tile_origins</cite>, ntiles[0] x ntiles[1] x 2 matrix of tile origin</dt><dd><p>coordinates</p>
</dd>
<dt>numpy.ndarray: <cite>tile_centers</cite>, ntiles[0] x ntiles[1] x 2 matrix of tile center</dt><dd><p>coordinates</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.find_tile_order">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">find_tile_order</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">xy_stage_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'XYStage'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">z_stage_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'ZDrive'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.find_tile_order" title="Link to this definition">¶</a></dt>
<dd><p>Find the order of tiles in a multi-tile acquisition</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Acquisition prefix. If None, will use the one in ops.
Defaults to None.</p></li>
<li><p><strong>xy_stage_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the XY stage. Defaults to “XYStage”.</p></li>
<li><p><strong>z_stage_name</strong> (<em>str</em><em>, </em><em>optional</em>) – Name of the Z stage. If None, will not load Z
positions. Defaults to “ZDrive”.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – Print information about the number of tiles found.
Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Dictionary of tile order with tuple (roi, col, row) as key and acquisition</dt><dd><p>order (across all ROIs) as value.</p>
</dd>
</dl>
<p>pandas.DataFrame: DataFrame containing tile position information.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.find_tile_overlap">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">find_tile_overlap</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coor1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coor2</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.find_tile_overlap" title="Link to this definition">¶</a></dt>
<dd><p>Find the overlap between two tiles</p>
<p>If tile1 is the stack, the overlap can be accessed by:
tile1[overlap_tile_1[1]:overlap_tile_1[3], overlap_tile_1[0]:overlap_tile_1[2]]</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>rect1</strong> (<em>tuple</em>) – Rectangle coordinates (x0, y0, x1, y1)</p></li>
<li><p><strong>rect2</strong> (<em>tuple</em>) – Rectangle coordinates (x0, y0, x1, y1)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Overlap in global coordinates (x0, y0, x1, y1)
tuple: Overlap in tile 1 (x0, y0, x1, y1)
tuple: Overlap in tile 2 (x0, y0, x1, y1)</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.get_tform_to_ref">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">get_tform_to_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">corrected_shifts</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.get_tform_to_ref" title="Link to this definition">¶</a></dt>
<dd><p>Load the transformation to reference for a tile</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition prefix</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – (roi, tileX, tileY) tuple</p></li>
<li><p><strong>corrected_shifts</strong> (<em>str</em><em>, </em><em>optional</em>) – Method used to correct shifts to reference.
If None, will use the one in <cite>ops</cite>. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with the transformation parameters</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.get_tile_corners">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">get_tile_corners</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.get_tile_corners" title="Link to this definition">¶</a></dt>
<dd><p>Find the corners of all tiles for a roi</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition prefix. For round-based acquisition, round 1 will be
used</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Roi ID</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt><cite>tile_corners</cite>, ntiles[0] x ntiles[1] x 2 x 4 matrix of tile</dt><dd><p>corners coordinates. Corners are in this order:
[(origin), (0, 1), (1, 1), (1, 0)]</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.load_tile_ref_coors">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">load_tile_ref_coors</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.load_tile_ref_coors" title="Link to this definition">¶</a></dt>
<dd><p>Load one single tile in the reference coordinates</p>
<p>This load a tile of <cite>prefix</cite> with channels/rounds registered</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>tile_coordinates</strong> (<em>tuple</em>) – (Roi, tileX, tileY) tuple</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition to load. If <cite>genes_round</cite> or <cite>barcode_round</cite> will load
all the rounds.</p></li>
<li><p><strong>filter_r</strong> (<em>bool</em><em>, </em><em>optional</em>) – Apply filter on rounds data? Parameters will be read
from <cite>ops</cite>. Default to True</p></li>
<li><p><strong>projection</strong> (<em>str</em><em>, </em><em>optional</em>) – Projection to load. If None, will use the one in
<cite>ops</cite>. Default to None</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Apply illumination correction. Default
to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A (X x Y x Nchannels x Nrounds) registered stack
np.array: A (X x Y) boolean array of bad pixels that fall outside image after</p>
<blockquote>
<div><p>registration</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.register_adjacent_tiles">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">register_adjacent_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_coors</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_ch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round_1_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlap_ratio</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.01</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debug</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.register_adjacent_tiles" title="Link to this definition">¶</a></dt>
<dd><p>Estimate shift between adjacent imaging tiles using phase correlation.</p>
<p>Shifts are typically very similar between different tiles, using shifts
estimated using a reference tile for the whole acquisition works well.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – path to image stacks.</p></li>
<li><p><strong>ref_coors</strong> (<em>tuple</em><em>, </em><em>optional</em>) – coordinates of the reference tile to use for
registration. Must not be along the bottom or right edge of image. If <cite>None</cite>
use <cite>ops[‘ref_tile’]</cite>. Defaults to None.</p></li>
<li><p><strong>ref_ch</strong> (<em>int</em><em>, </em><em>optional</em>) – reference channel used for registration. Defaults to 0.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – File name suffix. Defaults to ‘proj’.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Full name of the acquisition folder</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Remove black levels and correct
illumination before registration if True, return raw data otherwise. Default
to False</p></li>
<li><p><strong>overlap_ratio</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum overlap between masks to consider the
correlation results. Defaults to 0.01.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, print warnings when shifts are large.
Defaults to True.</p></li>
<li><p><strong>debug</strong> (<em>bool</em><em>, </em><em>optional</em>) – Return additional information for debugging. Defaults to
False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><cite>shift_right</cite>, X and Y shifts between different columns
numpy.array: <cite>shift_down</cite>, X and Y shifts between different rows
numpy.array: shape of the tile</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.register_all_rois_within">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">register_all_rois_within</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_ch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max-median'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi2use</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dimension_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.register_all_rois_within" title="Link to this definition">¶</a></dt>
<dd><p>Register all tiles within each ROI</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of acquisition to register. If None, will use the
one in <cite>ops</cite>. Defaults to None.</p></li>
<li><p><strong>ref_ch</strong> (<em>int</em><em>, </em><em>optional</em>) – Reference channel to use for registration. If None, will
use the one in <cite>ops</cite>. Defaults to None.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to use to load the images. Defaults to
‘max-median’.</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Correct illumination before registration.
Defaults to True.</p></li>
<li><p><strong>roi2use</strong> (<em>list</em><em>, </em><em>optional</em>) – List of ROI to use. If None or empty, will process all
ROIs. Defaults to None</p></li>
<li><p><strong>reload</strong> (<em>bool</em><em>, </em><em>optional</em>) – Reload saved shifts if True. Defaults to False.</p></li>
<li><p><strong>save_plot</strong> (<em>bool</em><em>, </em><em>optional</em>) – Save diagnostic plot. Defaults to True.</p></li>
<li><p><strong>dimension_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to use to find ROI dimension. Used
only if the acquisition is an overview. Defaults to ‘reference_prefix’.</p></li>
<li><p><strong>verbose</strong> (<em>int</em><em>, </em><em>optional</em>) – Verbosity level. Defaults to 1.</p></li>
<li><p><strong>use_slurm</strong> (<em>bool</em><em>, </em><em>optional</em>) – Use SLURM to parallelize the registration. Defaults
to True.</p></li>
<li><p><strong>job_dependency</strong> (<em>list</em><em>, </em><em>optional</em>) – List of job dependencies. Defaults to None.</p></li>
<li><p><strong>script_names</strong> (<em>str</em><em>, </em><em>optional</em>) – Script names for slurm jobs. Defaults to None.</p></li>
<li><p><strong>slurm_folder</strong> (<em>str</em><em>, </em><em>optional</em>) – Folder to save SLURM logs. Defaults to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of outputs from <cite>register_within_acquisition</cite></p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.register_within_acquisition">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">register_within_acquisition</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_ch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reload</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_plot</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dimension_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'genes_round_1_1'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_corrcoef</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.6</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">max_delta_shift</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">20</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">raise_on_empty_line</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_slurm</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dependency_type</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">job_dependency</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_folder</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">scripts_name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">slurm_options</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_names</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">batch_param_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.register_within_acquisition" title="Link to this definition">¶</a></dt>
<dd><p>Estimate shifts between all adjacent tiles of an roi</p>
<p>Saves shifts as <cite>reg/f”{prefix}_within”/f”{prefix}_{roi}_shifts.npz”</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – path to image stacks.</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – id of ROI to load.</p></li>
<li><p><strong>prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Full name of the acquisition folder.</p></li>
<li><p><strong>ref_ch</strong> (<em>int</em><em>, </em><em>optional</em>) – reference channel used for registration. Defaults to 0.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – File name suffix. Defaults to ‘proj’.</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Remove black levels and correct
illumination before registration if True, return raw data otherwise. Default
to False</p></li>
<li><p><strong>reload</strong> (<em>bool</em><em>, </em><em>optional</em>) – If target file already exists, reload instead of
recomputing. Defaults to True</p></li>
<li><p><strong>save_plot</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True save diagnostic plot. Defaults to False</p></li>
<li><p><strong>dimension_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix to use to find ROI dimension. Used
only if the acquisition is an overview. Defaults to ‘genes_round_1_1’</p></li>
<li><p><strong>min_corrcoef</strong> (<em>float</em><em>, </em><em>optional</em>) – Minimum correlation coefficient to consider a
shift as valid. Defaults to 0.6.</p></li>
<li><p><strong>max_delta_shift</strong> (<em>int</em><em>, </em><em>optional</em>) – Maximum shift, relative to median of the row or
column, to consider a shift as valid. Defaults to 20.</p></li>
<li><p><strong>verbose</strong> (<em>int</em><em>, </em><em>optional</em>) – Verbosity level. Defaults to 2.</p></li>
<li><p><strong>raise_on_empty_line</strong> (<em>bool</em><em>, </em><em>optional</em>) – Raise an error if a row or a column has no
valid shifts. If False, replace by the global median. Defaults to True</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>dictionary containing the shifts, tile shape and number of tiles</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.stitch_and_register">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">stitch_and_register</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reference_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">downsample</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_ch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_ch</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">estimate_scale</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">estimate_rotation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">use_masked_correlation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debug</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.stitch_and_register" title="Link to this definition">¶</a></dt>
<dd><p>Stitch target and reference stacks and align target to reference</p>
<p>To speed up registration, images are downsampled before estimating registration
parameters. These parameters are then applied to the full scale image.</p>
<p>The reference stack always use the “projection” from ops as suffix. The target uses
the same by default but that can be specified with <cite>target_suffix</cite></p>
<p>This does not use ops[‘max_shift_rounds’].</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data.</p></li>
<li><p><strong>reference_prefix</strong> (<em>str</em>) – Acquisition prefix to register the stitched image to.
Typically, “genes_round_1_1”.</p></li>
<li><p><strong>target_prefix</strong> (<em>str</em>) – Acquisition prefix to register.</p></li>
<li><p><strong>roi</strong> (<em>int</em><em>, </em><em>optional</em>) – ROI ID to register (as specified in MicroManager).
Defaults to 1.</p></li>
<li><p><strong>downsample</strong> (<em>int</em><em>, </em><em>optional</em>) – Downsample factor for estimating registration
parameter. Defaults to 5.</p></li>
<li><p><strong>ref_ch</strong> (<em>int</em><em>, </em><em>optional</em>) – Channel of the reference image used for registration.
Defaults to 0.</p></li>
<li><p><strong>target_ch</strong> (<em>int</em><em>, </em><em>optional</em>) – Channel of the target image used for registration.
Defaults to 0.</p></li>
<li><p><strong>estimate_scale</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to estimate scaling between target
and reference images. Defaults to False.</p></li>
<li><p><strong>estimate_rotation</strong> (<em>bool</em><em>, </em><em>optional</em>) – Whether to estimate rotation between target
and reference images. Defaults to True.</p></li>
<li><p><strong>target_suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – Suffix to use for target stack. If None, will use
the value from ops. Defaults to None.</p></li>
<li><p><strong>use_masked_correlation</strong> (<em>bool</em><em>, </em><em>optional</em>) – Use masked correlation for registration
Defaults to False.</p></li>
<li><p><strong>debug</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, return full xcorr. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Stitched target image after registration.
numpy.ndarray: Stitched reference image.
float: Estimate rotation angle.
tuple: Estimated X and Y shifts.
float: Estimated scaling factor.
dict: Debug information if <cite>debug</cite> is True.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.stitch_registered">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">stitch_registered</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">channels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ref_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">projection</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.stitch_registered" title="Link to this definition">¶</a></dt>
<dd><p>Load registered stack and stitch them</p>
<p>The output is in the reference coordinate.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Prefix of acquisition to stitch</p></li>
<li><p><strong>roi</strong> (<em>int</em>) – Roi ID</p></li>
<li><p><strong>channels</strong> (<em>list</em><em> or </em><em>int</em><em>, </em><em>optional</em>) – Channel id(s). Defaults to 0.</p></li>
<li><p><strong>ref_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – Prefix of reference acquisition to load shifts. If
None, load from ops. Defaults to None.</p></li>
<li><p><strong>filter_r</strong> (<em>bool</em><em>, </em><em>optional</em>) – Filter image before stitching? Defaults to False.</p></li>
<li><p><strong>projection</strong> (<em>str</em><em>, </em><em>optional</em>) – Projection to load. If None, will use the one in
<cite>ops</cite>. Default to None</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Correct illumination before stitching.
Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>stitched stack</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.stitch_tiles">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">stitch_tiles</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">roi</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'max'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ich</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correct_illumination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">shifts_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">register_channels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">allow_quick_estimate</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_r</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.stitch_tiles" title="Link to this definition">¶</a></dt>
<dd><p>Load and stitch tile images using saved tile shifts.</p>
<p>This will load the tile shifts saved by <cite>register_within_acquisition</cite></p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>data_path</strong> (<em>str</em>) – path to image stacks.</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – prefix specifying which images to load, e.g. ‘round_01_1’</p></li>
<li><p><strong>roi</strong> (<em>int</em><em>, </em><em>optional</em>) – id of ROI to load. Defaults to 1.</p></li>
<li><p><strong>suffix</strong> (<em>str</em><em>, </em><em>optional</em>) – filename suffix. Defaults to ‘fstack’.</p></li>
<li><p><strong>ich</strong> (<em>int</em><em>, </em><em>optional</em>) – index of the channel to stitch. Defaults to 0.</p></li>
<li><p><strong>correct_illumination</strong> (<em>bool</em><em>, </em><em>optional</em>) – Remove black levels and correct
illumination if True, return raw data otherwise. Default to False</p></li>
<li><p><strong>shifts_prefix</strong> (<em>str</em><em>, </em><em>optional</em>) – prefix to use to load tile shifts. If not
provided, use <cite>prefix</cite>. Defaults to None.</p></li>
<li><p><strong>register_channels</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, register channels before stitching.
Defaults to True.</p></li>
<li><p><strong>allow_quick_estimate</strong> (<em>bool</em><em>, </em><em>optional</em>) – If True, will estimate shifts from a
single tile if shifts.npz is not found. Defaults to False.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>stitched image.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>numpy.ndarray</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="iss_preprocess.pipeline.stitch.warp_stack_to_ref">
<span class="sig-prename descclassname"><span class="pre">iss_preprocess.pipeline.stitch.</span></span><span class="sig-name descname"><span class="pre">warp_stack_to_ref</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">stack</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prefix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tile_coors</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">interpolation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">bad_pixels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#iss_preprocess.pipeline.stitch.warp_stack_to_ref" title="Link to this definition">¶</a></dt>
<dd><p>Warp a stack to the reference coordinates</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>stack</strong> (<em>np.array</em>) – A (X x Y x Nchannels x Nrounds) stack</p></li>
<li><p><strong>data_path</strong> (<em>str</em>) – Relative path to data</p></li>
<li><p><strong>prefix</strong> (<em>str</em>) – Acquisition to use to find registration parameters</p></li>
<li><p><strong>tile_coors</strong> (<em>tuple</em>) – (Roi, tileX, tileY) tuple</p></li>
<li><p><strong>interpolation</strong> (<em>int</em><em>, </em><em>optional</em>) – Interpolation order. Defaults to 1.</p></li>
<li><p><strong>bad_pixels</strong> (<em>np.array</em><em>, </em><em>optional</em>) – A (X x Y) boolean array of bad pixels that fall
outside image after registration. If None, will not apply any mask. Defaults
to None.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A (X x Y x Nchannels x Nrounds) registered stack
np.array: A (X x Y) boolean array of bad pixels that fall outside image after</p>
<blockquote>
<div><p>registration</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>np.array</p>
</dd>
</dl>
</dd></dl>

</section>
<section id="module-iss_preprocess.pipeline">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-iss_preprocess.pipeline" title="Link to this heading">¶</a></h2>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="iss_preprocess.reg.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">iss_preprocess.reg package</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="iss_preprocess.io.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">iss_preprocess.io package</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2023, Znamenskiy lab
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">iss_preprocess.pipeline package</a><ul>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.ara_registration">iss_preprocess.pipeline.ara_registration module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.check_reg"><code class="docutils literal notranslate"><span class="pre">check_reg()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.crop_overview_registration"><code class="docutils literal notranslate"><span class="pre">crop_overview_registration()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.find_roi_position_on_cryostat"><code class="docutils literal notranslate"><span class="pre">find_roi_position_on_cryostat()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.load_coordinate_image"><code class="docutils literal notranslate"><span class="pre">load_coordinate_image()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.load_registration_reference"><code class="docutils literal notranslate"><span class="pre">load_registration_reference()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.load_registration_reference_metadata"><code class="docutils literal notranslate"><span class="pre">load_registration_reference_metadata()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.make_area_image"><code class="docutils literal notranslate"><span class="pre">make_area_image()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.overview_single_roi"><code class="docutils literal notranslate"><span class="pre">overview_single_roi()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.register_overview_to_reference"><code class="docutils literal notranslate"><span class="pre">register_overview_to_reference()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.ara_registration.spots_ara_infos"><code class="docutils literal notranslate"><span class="pre">spots_ara_infos()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.hybridisation">iss_preprocess.pipeline.hybridisation module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.estimate_channel_correction_hybridisation"><code class="docutils literal notranslate"><span class="pre">estimate_channel_correction_hybridisation()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.extract_hyb_spots_all"><code class="docutils literal notranslate"><span class="pre">extract_hyb_spots_all()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.extract_hyb_spots_roi"><code class="docutils literal notranslate"><span class="pre">extract_hyb_spots_roi()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.extract_hyb_spots_tile"><code class="docutils literal notranslate"><span class="pre">extract_hyb_spots_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.hyb_spot_cluster_means"><code class="docutils literal notranslate"><span class="pre">hyb_spot_cluster_means()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.load_and_register_hyb_tile"><code class="docutils literal notranslate"><span class="pre">load_and_register_hyb_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.hybridisation.setup_hyb_spot_calling"><code class="docutils literal notranslate"><span class="pre">setup_hyb_spot_calling()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.pipeline">iss_preprocess.pipeline.pipeline module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.call_spots"><code class="docutils literal notranslate"><span class="pre">call_spots()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.correct_shifts"><code class="docutils literal notranslate"><span class="pre">correct_shifts()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.create_all_single_averages"><code class="docutils literal notranslate"><span class="pre">create_all_single_averages()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.create_grand_averages"><code class="docutils literal notranslate"><span class="pre">create_grand_averages()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.create_single_average"><code class="docutils literal notranslate"><span class="pre">create_single_average()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.overview_for_ara_registration"><code class="docutils literal notranslate"><span class="pre">overview_for_ara_registration()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.project_and_average"><code class="docutils literal notranslate"><span class="pre">project_and_average()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.register_acquisition"><code class="docutils literal notranslate"><span class="pre">register_acquisition()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.register_reference_tile"><code class="docutils literal notranslate"><span class="pre">register_reference_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.segment_and_stitch_mcherry_cells"><code class="docutils literal notranslate"><span class="pre">segment_and_stitch_mcherry_cells()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.pipeline.setup_channel_correction"><code class="docutils literal notranslate"><span class="pre">setup_channel_correction()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.project">iss_preprocess.pipeline.project module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.check_projection"><code class="docutils literal notranslate"><span class="pre">check_projection()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.check_roi_dims"><code class="docutils literal notranslate"><span class="pre">check_roi_dims()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.project_round"><code class="docutils literal notranslate"><span class="pre">project_round()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.project_tile"><code class="docutils literal notranslate"><span class="pre">project_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.project_tile_by_coors"><code class="docutils literal notranslate"><span class="pre">project_tile_by_coors()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.project_tile_row"><code class="docutils literal notranslate"><span class="pre">project_tile_row()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.project.reproject_failed"><code class="docutils literal notranslate"><span class="pre">reproject_failed()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.register">iss_preprocess.pipeline.register module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.correct_hyb_shifts"><code class="docutils literal notranslate"><span class="pre">correct_hyb_shifts()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.correct_shifts_roi"><code class="docutils literal notranslate"><span class="pre">correct_shifts_roi()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.correct_shifts_single_round_roi"><code class="docutils literal notranslate"><span class="pre">correct_shifts_single_round_roi()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.correct_shifts_to_ref"><code class="docutils literal notranslate"><span class="pre">correct_shifts_to_ref()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.estimate_shifts_by_coors"><code class="docutils literal notranslate"><span class="pre">estimate_shifts_by_coors()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.filter_ransac_shifts"><code class="docutils literal notranslate"><span class="pre">filter_ransac_shifts()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.load_and_register_raw_stack"><code class="docutils literal notranslate"><span class="pre">load_and_register_raw_stack()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.load_and_register_sequencing_tile"><code class="docutils literal notranslate"><span class="pre">load_and_register_sequencing_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.load_and_register_tile"><code class="docutils literal notranslate"><span class="pre">load_and_register_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.merge_shifts"><code class="docutils literal notranslate"><span class="pre">merge_shifts()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.register_channels_by_pairs"><code class="docutils literal notranslate"><span class="pre">register_channels_by_pairs()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.register_fluorescent_tile"><code class="docutils literal notranslate"><span class="pre">register_fluorescent_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.run_correct_shifts"><code class="docutils literal notranslate"><span class="pre">run_correct_shifts()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.register.run_register_reference_tile"><code class="docutils literal notranslate"><span class="pre">run_register_reference_tile()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.segment">iss_preprocess.pipeline.segment module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.add_mask_id"><code class="docutils literal notranslate"><span class="pre">add_mask_id()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.filter_mcherry_cells"><code class="docutils literal notranslate"><span class="pre">filter_mcherry_cells()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.find_edge_touching_masks"><code class="docutils literal notranslate"><span class="pre">find_edge_touching_masks()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.get_big_masks"><code class="docutils literal notranslate"><span class="pre">get_big_masks()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.get_cell_masks"><code class="docutils literal notranslate"><span class="pre">get_cell_masks()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.get_overlap_regions"><code class="docutils literal notranslate"><span class="pre">get_overlap_regions()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.get_stack_for_cellpose"><code class="docutils literal notranslate"><span class="pre">get_stack_for_cellpose()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.make_cell_dataframe"><code class="docutils literal notranslate"><span class="pre">make_cell_dataframe()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.remove_all_duplicate_masks"><code class="docutils literal notranslate"><span class="pre">remove_all_duplicate_masks()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.run_cellpose_segmentation"><code class="docutils literal notranslate"><span class="pre">run_cellpose_segmentation()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.run_mask_projection"><code class="docutils literal notranslate"><span class="pre">run_mask_projection()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.save_curated_dataframes"><code class="docutils literal notranslate"><span class="pre">save_curated_dataframes()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.save_mcherry_mask_df"><code class="docutils literal notranslate"><span class="pre">save_mcherry_mask_df()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.save_unmixing_coefficients"><code class="docutils literal notranslate"><span class="pre">save_unmixing_coefficients()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.segment_all_rois"><code class="docutils literal notranslate"><span class="pre">segment_all_rois()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.segment_all_tiles"><code class="docutils literal notranslate"><span class="pre">segment_all_tiles()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.segment_mcherry_tile"><code class="docutils literal notranslate"><span class="pre">segment_mcherry_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.segment_roi"><code class="docutils literal notranslate"><span class="pre">segment_roi()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.segment_spots"><code class="docutils literal notranslate"><span class="pre">segment_spots()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.segment.unmix_tile"><code class="docutils literal notranslate"><span class="pre">unmix_tile()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.sequencing">iss_preprocess.pipeline.sequencing module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.basecall_tile"><code class="docutils literal notranslate"><span class="pre">basecall_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.compute_spot_sign_image"><code class="docutils literal notranslate"><span class="pre">compute_spot_sign_image()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.detect_genes_on_tile"><code class="docutils literal notranslate"><span class="pre">detect_genes_on_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.estimate_channel_correction"><code class="docutils literal notranslate"><span class="pre">estimate_channel_correction()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.get_reference_spots"><code class="docutils literal notranslate"><span class="pre">get_reference_spots()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.load_spot_sign_image"><code class="docutils literal notranslate"><span class="pre">load_spot_sign_image()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.run_omp_on_tile"><code class="docutils literal notranslate"><span class="pre">run_omp_on_tile()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.setup_barcode_calling"><code class="docutils literal notranslate"><span class="pre">setup_barcode_calling()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.sequencing.setup_omp"><code class="docutils literal notranslate"><span class="pre">setup_omp()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline.stitch">iss_preprocess.pipeline.stitch module</a><ul>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.calculate_tile_positions"><code class="docutils literal notranslate"><span class="pre">calculate_tile_positions()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.find_tile_order"><code class="docutils literal notranslate"><span class="pre">find_tile_order()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.find_tile_overlap"><code class="docutils literal notranslate"><span class="pre">find_tile_overlap()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.get_tform_to_ref"><code class="docutils literal notranslate"><span class="pre">get_tform_to_ref()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.get_tile_corners"><code class="docutils literal notranslate"><span class="pre">get_tile_corners()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.load_tile_ref_coors"><code class="docutils literal notranslate"><span class="pre">load_tile_ref_coors()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.register_adjacent_tiles"><code class="docutils literal notranslate"><span class="pre">register_adjacent_tiles()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.register_all_rois_within"><code class="docutils literal notranslate"><span class="pre">register_all_rois_within()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.register_within_acquisition"><code class="docutils literal notranslate"><span class="pre">register_within_acquisition()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.stitch_and_register"><code class="docutils literal notranslate"><span class="pre">stitch_and_register()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.stitch_registered"><code class="docutils literal notranslate"><span class="pre">stitch_registered()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.stitch_tiles"><code class="docutils literal notranslate"><span class="pre">stitch_tiles()</span></code></a></li>
<li><a class="reference internal" href="#iss_preprocess.pipeline.stitch.warp_stack_to_ref"><code class="docutils literal notranslate"><span class="pre">warp_stack_to_ref()</span></code></a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-iss_preprocess.pipeline">Module contents</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script src="_static/documentation_options.js?v=37f418d5"></script>
    <script src="_static/doctools.js?v=9bcbadda"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/scripts/furo.js?v=5fa4622c"></script>
    </body>
</html>